File: python/p4.py
def isPalindrome(num):
	return str(num) == str(num)[::-1]

def largestPalindrome(numDigits):
	maxNum = int('9'*numDigits)
	minNum = int('1'+'0'*max(0, numDigits-1))
	factors = [i for i in xrange(minNum, maxNum)]
	maxPalindrome = 0
	for i in factors:
		for j in factors:
			prod = i*j
			if isPalindrome(prod) and prod > maxPalindrome:
				maxPalindrome = prod
	return maxPalindrome

# Return the largest palindrome that is a product of two 3-digit numbers
print largestPalindrome(3)

File: javascript/p4.js
const isPalindrome = (num) => {
	return num.toString().split("").reverse().join("") === num.toString();
};

const largestPalindrome = (numDigits) => {
	let maxNum = parseInt('9'.repeat(numDigits));
	let minNum = parseInt('1' + '0'.repeat(numDigits-1));
	let palindrome = []

	factors = [...Array(maxNum-minNum+1).keys()].map((n) => (n+minNum));

	for (i of factors) {
		for (j of factors) {
			prod = i * j;
			if (isPalindrome(prod)) {
				palindrome.push(prod);
			}
		}
	}

	return Math.max(...palindrome);
};

// Return the largest palindrome that is a product of two 3-digit numbers
console.log(largestPalindrome(3));